File: bindings/python/native/iota_client/_high_level_api.py
from iota_client._base_api import BaseAPI


class HighLevelAPI(BaseAPI):

    def get_outputs(self, output_ids):
        """Fetch OutputResponse from provided OutputIds (requests are sent in parallel).
        """
        return self.send_message('getOutputs', {
            'outputIds': output_ids
        })

    def try_get_outputs(self, output_ids):
        """Try to get OutputResponse from provided OutputIds.
           Requests are sent in parallel and errors are ignored, can be useful for spent outputs.
        """
        return self.send_message('tryGetOutputs', {
            'outputIds': output_ids
        })

    def find_blocks(self, block_ids):
        """Find all blocks by provided block IDs.
        """
        return self.send_message('findBlocks', {
            'blockIds': block_ids
        })

    def retry(self, block_id):
        """Retries (promotes or reattaches) a block for provided block id. Block should only be
           retried only if they are valid and haven't been confirmed for a while.
        """
        return self.send_message('retry', {'blockId': block_id})

    def retry_until_included(self, block_id, interval=None, max_attempts=None):
        """Retries (promotes or reattaches) a block for provided block id until it's included (referenced by a
           milestone). Default interval is 5 seconds and max attempts is 40. Returns the included block at first
           position and additional reattached blocks.
        """
        return self.send_message('retryUntilIncluded', {
            'blockId': block_id,
            'interval': interval,
            'maxAttempts': max_attempts
        })

    def consolidate_funds(self, secret_manager, generate_addresses_options):
        """Function to consolidate all funds from a range of addresses to the address with the lowest index in that range
           Returns the address to which the funds got consolidated, if any were available.
        """
        return self.send_message('consolidateFunds', {
            'secretManager': secret_manager,
            'generateAddressesOptions': generate_addresses_options,
        })

    def find_inputs(self, addresses, amount):
        """Function to find inputs from addresses for a provided amount (useful for offline signing)
        """
        return self.send_message('findInputs', {
            'addresses': addresses,
            'amount': amount
        })

    def find_outputs(self, output_ids, addresses):
        """Find all outputs based on the requests criteria. This method will try to query multiple nodes if
           the request amount exceeds individual node limit.
        """
        return self.send_message('findOutputs', {
            'outputIds': output_ids,
            'addresses': addresses
        })

    def reattach(self, block_id):
        """Reattaches blocks for provided block id. Blocks can be reattached only if they are valid and haven't been
           confirmed for a while.
        """
        return self.send_message('reattach', {
            'blockId': block_id
        })

    def reattach_unchecked(self, block_id):
        """Reattach a block without checking if it should be reattached.
        """
        return self.send_message('reattachUnchecked', {
            'blockId': block_id
        })

    def promote(self, block_id):
        """Promotes a block. The method should validate if a promotion is necessary through get_block. If not, the
           method should error out and should not allow unnecessary promotions.
        """
        return self.send_message('promote', {
            'blockId': block_id
        })

    def promote_unchecked(self, block_id):
        """Promote a block without checking if it should be promoted.
        """
        return self.send_message('promoteUnchecked', {
            'blockId': block_id
        })

File: bindings/java/iota-client-java/src/main/java/org/iota/apis/HighLevelApi.java
// Copyright 2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package org.iota.apis;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.iota.types.*;
import org.iota.types.ids.BlockId;
import org.iota.types.ids.OutputId;
import org.iota.types.secret.GenerateAddressesOptions;
import org.iota.types.secret.Range;
import org.iota.types.secret.SecretManager;

import java.util.AbstractMap.SimpleEntry;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class HighLevelApi extends BaseApi {

    public HighLevelApi(ClientConfig clientConfig) {
        super(clientConfig);
    }

    public List<Map.Entry<Output, OutputMetadata>> getOutputs(OutputId[] outputIds) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("outputIds", JsonUtils.toJson(outputIds));

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("GetOutputs", o));

        List<Map.Entry<Output, OutputMetadata>> outputs = new ArrayList<>();
        for (int i = 0; i < responsePayload.size(); i++) {
            Output output = new Output(responsePayload.get(i).getAsJsonObject().get("output").getAsJsonObject());
            OutputMetadata metadata = new OutputMetadata(responsePayload.get(i).getAsJsonObject().get("metadata").getAsJsonObject());
            outputs.add(new SimpleEntry(output, metadata));
        }

        return outputs;
    }

    public List<Map.Entry<Output, OutputMetadata>> tryGetOutputs(OutputId[] outputIds) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("outputIds", JsonUtils.toJson(outputIds));

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("TryGetOutputs", o));

        List<Map.Entry<Output, OutputMetadata>> outputs = new ArrayList<>();
        for (int i = 0; i < responsePayload.size(); i++) {
            Output output = new Output(responsePayload.get(i).getAsJsonObject().get("output").getAsJsonObject());
            OutputMetadata metadata = new OutputMetadata(responsePayload.get(i).getAsJsonObject().get("metadata").getAsJsonObject());
            outputs.add(new SimpleEntry(output, metadata));
        }

        return outputs;
    }

    public Block[] findBlocks(BlockId[] blockIds) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("blockIds", JsonUtils.toJson(blockIds));

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("FindBlocks", o));

        Block[] blocks = new Block[responsePayload.size()];
        for (int i = 0; i < responsePayload.size(); i++)
            blocks[i] = new Block(responsePayload.get(i).getAsJsonObject());

        return blocks;
    }

    public Map.Entry<BlockId, Block> retry(BlockId blockId) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("Retry", o));

        return Map.entry(new BlockId(responsePayload.get(0).getAsString()), new Block(responsePayload.get(1).getAsJsonObject()));
    }

    public LinkedHashMap<BlockId, Block> retryUntilIncluded(BlockId blockId, int interval, int maxAttempts) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());
        o.addProperty("interval", interval);
        o.addProperty("maxAttempts", maxAttempts);

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("RetryUntilIncluded", o));

        LinkedHashMap<BlockId, Block> blocks = new LinkedHashMap<BlockId, Block>();
        for (JsonElement entry : responsePayload) {
            JsonArray e = entry.getAsJsonArray();
            blocks.put(new BlockId(e.get(0).getAsString()), new Block(e.get(1).getAsJsonObject()));
        }

        return blocks;
    }

    public String consolidateFunds(SecretManager secretManager, GenerateAddressesOptions generateAddressesOptions) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("secretManager", secretManager.getJson());
        o.add("generateAddressesOptions", generateAddressesOptions.getJson());

        String responsePayload = callBaseApi(new ClientCommand("ConsolidateFunds", o)).getAsString();

        return responsePayload;
    }

    public UtxoInput[] findInputs(String[] addresses, int amount) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("addresses", JsonUtils.toJson(addresses));
        o.addProperty("amount", amount);

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("FindInputs", o));

        UtxoInput[] inputs = new UtxoInput[responsePayload.size()];
        for (int i = 0; i < responsePayload.size(); i++)
            inputs[i] = new UtxoInput(responsePayload.get(i).getAsJsonObject());

        return inputs;
    }

    public List<Map.Entry<Output, OutputMetadata>> findOutputs(OutputId[] outputIds, String[] addresses) throws ClientException {
        JsonObject o = new JsonObject();
        o.add("outputIds", JsonUtils.toJson(outputIds));
        o.add("addresses", JsonUtils.toJson(addresses));

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("FindOutputs", o));

        List<Map.Entry<Output, OutputMetadata>> outputs = new ArrayList<>();
        for (int i = 0; i < responsePayload.size(); i++) {
            Output output = new Output(responsePayload.get(i).getAsJsonObject().get("output").getAsJsonObject());
            OutputMetadata metadata = new OutputMetadata(responsePayload.get(i).getAsJsonObject().get("metadata").getAsJsonObject());
            outputs.add(new SimpleEntry(output, metadata));
        }

        return outputs;
    }

    public Map.Entry<BlockId, Block> reattach(BlockId blockId) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("Reattach", o));

        return Map.entry(new BlockId(responsePayload.get(0).getAsString()), new Block(responsePayload.get(1).getAsJsonObject()));
    }

    public Map.Entry<BlockId, Block> reattachUnchecked(BlockId blockId) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("ReattachUnchecked", o));

        return Map.entry(new BlockId(responsePayload.get(0).getAsString()), new Block(responsePayload.get(1).getAsJsonObject()));
    }

    public Map.Entry<BlockId, Block> promote(BlockId blockId) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("Promote", o));

        return Map.entry(new BlockId(responsePayload.get(0).getAsString()), new Block(responsePayload.get(1).getAsJsonObject()));
    }

    public Map.Entry<BlockId, Block> promoteUnchecked(BlockId blockId) throws ClientException {
        JsonObject o = new JsonObject();
        o.addProperty("blockId", blockId.toString());

        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("PromoteUnchecked", o));

        return Map.entry(new BlockId(responsePayload.get(0).getAsString()), new Block(responsePayload.get(1).getAsJsonObject()));
    }

}
